<?xml version="1.0"?>
<launch>
  <!-- All Arugements Definition -->
    <!-- Args of Tf Frames -->
    <!-- No need to change -->
    <arg name="base_frame_"   value="/base" />
    <arg name="laser_frame_" value="/laser" />
    <arg name="odom_frame_" value="/odom" />
    <arg name="pub_pose_topic_" value="pose" />
    <arg name="pub_laser_topic_" value="laser" />
    <arg name="pub_path_topic_" value="path" />

    <!-- Topics Arguments -->
    <!-- Change to the topic names in the rosbag or others -->
    <arg name="sub_laser_topic_"  value="/robot/front_laser/scan" />
    <arg name="sub_wheel_odom_topic_"  value="/robot/throttle/odom" />
    <!-- Other Predefined parameters -->
    <arg name="use_wheel_odom_" value="true" />
    <arg name="use_laser_" value="true" />
    <arg name="use_wheel_odom_prior_guess_" value="true" />
    <arg name="use_darko_cfg_" value="true" />

  <!-- Some static tf configuration -->
  <!-- Example: static_transform_publisher x y z yaw pitch roll frame_id child_frame_id period_in_ms -->
  <group if="$(arg use_darko_cfg_)">
    <!-- yaw,pitch,raw: [-0.785 0 3.142] to quanterion: [xyzw]=[ 0.92396 -0.3825 0 0 ] -->
    <node pkg="tf2_ros" type="static_transform_publisher" name="robot_to_sensor" args="0.421 -0.327 0.044 0.92396 -0.3825 0 0 $(arg base_frame_) $(arg laser_frame_)"/>
  </group>
  <group unless="$(arg use_darko_cfg_)">
    <node pkg="tf2_ros" type="static_transform_publisher" name="robot_to_sensor" args="0 0 0 0 0 0 0 $(arg base_frame_) $(arg laser_frame_)"/>
  </group>

  <!-- The node of slam  -->
  <node pkg="slam_demo" name="slam_node" type="slam_node" output="screen">
    <param name="use_darko_cfg_"   value="$(arg use_darko_cfg_)" />
    <param name="use_wheel_odom_prior_guess_" value="$(arg use_wheel_odom_prior_guess_)" />
    <param name="use_wheel_odom_" value="$(arg use_wheel_odom_)" />
    <param name="use_laser_" value="$(arg use_laser_)" />
    <param name="base_frame_"   value="$(arg base_frame_)" />
    <param name="laser_frame_" value="$(arg laser_frame_)" />
    <param name="odom_frame_" value="$(arg odom_frame_)" />
    <param name="sub_laser_topic_" value="$(arg sub_laser_topic_)"/>
    <param name="sub_wheel_odom_topic_" value="$(arg sub_wheel_odom_topic_)"/>
  </node>

</launch>
  <!-- <node pkg="rviz" type="rviz" name="rviz" args="-d $(find slam_demo)/rviz/odom.rviz"/> -->
  <!-- <node pkg="rosbag" type="play" name="rosbag" args="/home/ros/catkin_ws/src/darko/darko_ros_bag/acg/2022-11-28-17-40-18-robolab.bag
  topics /robot/front_laser/scan /robot/throttle/odom -r5" /> -->
